clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1256,65 +1182,331 @@ static void populateIntHeader(SYCLIntegrationHeader &H, const StringRef Name,
+  // TODO: Not sure what the 'fully qualified type's purpose is here, the type
+  // itself should have its full qualified name, so figure out what the purpose
+  // is.
+  const TemplateArgumentList *TAL =
+      KernelCallerFunc->getTemplateSpecializationArgs();
+  return TypeName::getFullyQualifiedType(TAL->get(0).getAsType(), Ctx,
+                                         /*WithGlobalNSPrefix=*/true);
+}
+
+// Gets a name for the kernel caller func, calculated from the first template
+// argument.
+static std::string constructKernelName(Sema &S, FunctionDecl *KernelCallerFunc,
                                        MangleContext &MC) {
+  QualType KernelNameType =
+      calculateKernelNameType(S.getASTContext(), KernelCallerFunc);
+
+  if (S.getLangOpts().SYCLUnnamedLambda)
+    return PredefinedExpr::ComputeName(S.getASTContext(),
+                                       PredefinedExpr::UniqueStableNameType,
+                                       KernelNameType);
+
   SmallString<256> Result;
   llvm::raw_svector_ostream Out(Result);
 
   MC.mangleTypeName(KernelNameType, Out);
   return std::string(Out.str());
 }
 
-static FunctionDecl *
-CreateOpenCLKernelDeclaration(ASTContext &Context, StringRef Name,
-                              ArrayRef<ParamDesc> ParamDescs) {
-
-  DeclContext *DC = Context.getTranslationUnitDecl();
-  QualType RetTy = Context.VoidTy;
-  SmallVector<QualType, 8> ArgTys;
-
-  // Extract argument types from the descriptor array:
-  std::transform(
-      ParamDescs.begin(), ParamDescs.end(), std::back_inserter(ArgTys),
-      [](const ParamDesc &PD) -> QualType { return std::get<0>(PD); });
-  FunctionProtoType::ExtProtoInfo Info(CC_OpenCLKernel);
-  QualType FuncTy = Context.getFunctionType(RetTy, ArgTys, Info);
-  DeclarationName DN = DeclarationName(&Context.Idents.get(Name));
-
-  FunctionDecl *OpenCLKernel = FunctionDecl::Create(
-      Context, DC, SourceLocation(), SourceLocation(), DN, FuncTy,
-      Context.getTrivialTypeSourceInfo(RetTy), SC_None);
-
-  llvm::SmallVector<ParmVarDecl *, 16> Params;
-  int i = 0;
-  for (const auto &PD : ParamDescs) {
-    auto P = ParmVarDecl::Create(Context, OpenCLKernel, SourceLocation(),
-                                 SourceLocation(), std::get<1>(PD),
-                                 std::get<0>(PD), std::get<2>(PD), SC_None, 0);
-    P->setScopeInfo(0, i++);
-    P->setIsUsed();
-    Params.push_back(P);
+// anonymous namespace so these don't get linkage.
+namespace {
+template <typename... Handlers>
+static void VisitKernelFields(RecordDecl::field_range Fields,
+                              Handlers &... handlers) {
+
+// Implements the 'for-each-visitor'  pattern.
+#define KF_FOR_EACH(FUNC)                                                      \
+  (void)std::initializer_list<int> { (handlers.FUNC(Field, ArgTy), 0)... }
+
+  for (const auto &Field : Fields) {
+    QualType ArgTy = Field->getType();
```

Comment:
I prefer call this thing not `ArgTy` but `KernelObjFieldTy` or something like this... Because I was confused by it, I was thinking that it is type of kernel argument for some time.

Reply:
Well, the idea is that in this function it is a 'generic' function, so all it cares about is the field.  Perhaps FieldTy?

Reply:
Ok. Maybe we also need to rename function so the name of it won't contain `Kernel` at all. I've seen this function is used for some structs, not only SYCL kernel object.

Reply:
Yep, thats a good idea.  It originally ONLY did the kernel fields :) I discovered the struct recursion later.